26.md

```diff
@@ -0,0 +1,80 @@
+NIP: 26
+=======
+
+Delegated Event Signing
+-----
+
+`draft` `mandatory` `author:markharding` `author:minds`
+
+This NIP defines how events should be verified and signed to support generating events on behalf of someone else. It should be possible to sign Nostr events from other keypairs.
+
+Another application of this proposal is to abstract away the use of the 'root' keypairs when interacting with clients. For example, a user could generate new keypairs for each client they wish to use and authorize those keypairs to generate events on behalf of their root pubkey, where the root keypair is stored in cold storage. 
+
+#### Introducing the 's' tag
+
+This NIP introduces a new tag: `s` which is formatted as follows:
+
+```json
+[
+  "s",
+  <signed pairing payload (base64 encoded)>,
+  <signed pairing signature (64-bytes shnnorr signature of the sha256 hash of the signed pairing payload>
+]
+```
+
+##### Signed Pairing Payload
+
+The Signed Pairing Payload should be a `base64` encoded JSON object as follows: 
```

Comment:
why base64 encode it? could it not just be a json string?

Reply:
yeah, it doesn't need to be base64 encoded, but stringified json is pretty igly imho

Reply:
well humans won't be consuming this, base64 encoded is 141 bytes and simple json string is 104. seems like a no brainer.

Reply:
I agree it is ugly, I think they should all go in the same tag array:

```
[
  "subkey",
  <signed pairing signature (64-bytes schnorr signature of the sha256 hash of the signed pairing payload>,
  <32-bytes hex-encoded public key of who is authorized to sign>,
  <unix timestamp of issued time>,
  <optional, if present unix timestamp of invalidation time>
]
```

77 bytes.

But I think we can remove the issued time and keep just the invalidation time.

Or just keep the 4th string for the [runes-like](https://github.com/rustyrussell/runes) thing @jb55 has in mind which is indeed pretty cool.

Reply:
I like fiatjaf's suggestion. In which case I suggest the signature to be done on the JSON of the following array `[<pubkey of delegate>,<issue time>[, <expiry>]]` (no whitespace)

Reply:
> Or just keep the 4th string for the runes-like thing @jb55 has in mind which is indeed pretty cool

yeah I think this might be the way to go, for now we can spec the expiry string as

`time<1659665936` - current time is before `Fri Aug  5 02:19:21 UTC 2022`

where the future restrictions spec could be:

`kind=1` kind is equal to 1
`kind!4` kind is not equal to 4

combined: `time<1659665936 & kind!4`

Reply:
> I like fiatjaf's suggestion. In which case I suggest the signature to be done on the JSON of the following array `[<pubkey of delegate>,<issue time>[, <expiry>]]` (no whitespace)

I'm not really a fan of not have key values here - its confusing for developer to understand and also makes it difficult to add or remove fields. For example, you have invalidation/expires_at being optional, but what if we want to include a 'scope' option which limits what a subkey can do? Is invalidation/expires_at supposed to be null?

Reply:
that's why I think `expiry` should just be `conditions` or `scope`, with the expiry condition set to `time<1659665936`. then we don't have to add a new field for every new condition and we can define this field in another NIP.

I'm fine with the object as well though if we think we would need to extend this further. Since [[String]] tags are so common on nostr it's not a problem to process it this way, we don't do json-encoded stuff in tags very often. I'm mostly indifferent though in terms of how it's encoded.

Reply:
@markharding Yeah, null is not allowed by NIP-01 so empty strings should be used instead.
@jb55's suggestion of using runes is a better approach and we won't need a key for each new condition. I also want to be able to search for events signed by a third party but if the pubkey is wrapped in an object of course it will be easier for developers to reason about it but at the expense of relays and clients to index/search for them.
If there's no intention on allowing clients to search for these events, then we should consider using `subkey` instead of `s` since tags with 1 letter are meant to be indexable and query-able.

I propose the following format: `["s", <delegate's pubkey in hex>, <scope>, <signed payload>]`
Where:
- `scope` follows [runes-like](https://github.com/rustyrussell/runes) spec
- `signed payload` is the signature of the JSON stringified tag `["s", <delegate's pubkey in hex>, <scope>]` without spaces

Issue time is NOT needed since you can include `created_at>1659665936` in `scope`, but if you really need it then it's okay, it should be added somewhere after the `delegate's pubkey`.

Here's some examples:
- Allow signing events of any kind by `aabb` created between `1659665936` and `1660665936`:
  `["s", "aabb", "created_at>1659665936 & created_at<1660665936", "...signature"]`
- Allow signing kind 1 events by `ccdd` created after `1659665936`:
  `["s", "ccdd", "kind=1 & created_at>1659665936", "...signature"]`
- Allow signing kind 1 events by `eeff` forever:
  `["s", "ccdd", "kind=1", "...signature"]`

Reply:
To revoke this permission, the owner of the delegated pubkey could send an event with tag having a capital `S`:
`["S", <delegate's pubkey>, <scope>, "<original signed payload>"]"`.
So if I want to revoke the permission of `ccdd` to send kind 1 events after `1659665936`, I'll include the following tag:
`["S", "ccdd", "kind=1 & created_at>1659665936", "original signature"]`

Reply:
With my suggestion it is trivial for existing clients/relays to search for all the delegations and revocations.

Reply:
@Cameri do you think it is worth implementing this same protocol here? https://github.com/rustyrussell/runes

Reply:
Yes, I think it's perfect for this use-case.

 That library may seem like a lot of code but it's bloated with comments and human-readable validation tests... but we could implement a much shorter compliant version that just returns true or false.